arrows/ffmpeg/ffmpeg_video_input.cxx

```diff
@@ -0,0 +1,832 @@
+bool
+ffmpeg_video_input
+::next_frame( kwiver::vital::timestamp& ts,
+              uint32_t timeout )
+{
+  if (!d->is_opened())
+  {
+    throw vital::file_not_read_exception(d->video_path, "Video not open");
+  }
+
+  bool ret = d->advance();
+  if (!ret && !d->end_of_video)
+  {
+    // If it's the first frame that fails
+    // -> first frame as the end of the video
+    d->number_of_frames_valid = true;
+  }
+
+  d->end_of_video = !ret;
+  if (ret)
+  {
+    ts = this->frame_timestamp();
+    if (!d->number_of_frames_valid)
+    {
+      d->number_of_frames = ts.get_frame();
+    }
+  };
+  return ret;
+}
+
+// ------------------------------------------------------------------
+bool ffmpeg_video_input::seek_frame(kwiver::vital::timestamp& ts,
+  kwiver::vital::timestamp::frame_t frame_number,
+  uint32_t timeout)
+{
+  // Quick return if the stream isn't open.
+  if (!d->is_opened())
+  {
+    throw vital::file_not_read_exception(d->video_path, "Video not open");
+    return false;
+  }
+  if (frame_number <= 0)
+  {
+    return false;
+  }
+
+  // Quick return if the stream isn't open.
+  if (timeout != 0)
+  {
+    LOG_WARN(this->logger(), "Timeout argument is not supported.");
+  }
+
+  int current_frame_number = this->frame_timestamp().get_frame();
+  // If current frame number is greater than requested frame reopen
+  // file to reset to start
+  if (current_frame_number > frame_number)
+  {
+    this->open(d->video_path); // Calls close on current video
+    current_frame_number = this->frame_timestamp().get_frame();
+  }
+
+  // Just advance video until the requested frame is reached
+  for (int i = current_frame_number; i < frame_number; ++i)
+  {
+    if (!this->next_frame(ts))
+    {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+
+// ------------------------------------------------------------------
+kwiver::vital::image_container_sptr
+ffmpeg_video_input
+::frame_image( )
+{
+  // Quick return if the stream isn't valid
+  if (!d->is_valid())
+  {
+    return nullptr;
+  }
+
+  AVCodecContext* enc = d->f_format_context->streams[d->f_video_index]->codec;
+
+  // If we have not already converted this frame, try to convert it
+  if (!d->current_image_memory && d->f_frame->data[0] != 0)
+  {
+    int width = enc->width;
+    int height = enc->height;
+    int num_pixels = 3;
+    vital::image_pixel_traits pixel_trait = vital::image_pixel_traits_of<unsigned char>();
+    bool direct_copy;
+
+    // If the pixel format is not recognized by then convert the data into RGB_24
+    switch (enc->pix_fmt)
+    {
+      case AV_PIX_FMT_GRAY8:
+      {
+        num_pixels = 1;
+        direct_copy = true;
+        break;
+      }
+      case AV_PIX_FMT_RGBA:
+      {
+        num_pixels = 4;
+        direct_copy = true;
+        break;
+      }
+      case AV_PIX_FMT_MONOWHITE:
+      case AV_PIX_FMT_MONOBLACK:
+      {
+        num_pixels = 1;
+        pixel_trait = vital::image_pixel_traits_of<bool>();
```

Comment:
I'm not sure if mono images are actually stored as an array of `bool` or or not.  It would be interesting to test.  I don't know that we have any test data though.